### PR TITLE
Qwen mrope cont

### PR DESCRIPTION
**Pull request by entrpn**
Original author @eitanporat in [#2727](https://github.com/AI-Hypercomputer/maxtext/pull/2727)

# Description

Implemented multi-modal rope and integrated in training and inference

# Tests

Compared to the pytorch implementation
`tests/multimodal_rope_check.py`

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [X] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [X] I have necessary comments in my code, particularly in hard-to-understand areas.
- [X] I have run end-to-end tests tests and provided workload links above if applicable.
- [X] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/MaxText/multimodal_utils.py [hunk 1]
```diff
@@ -775,6 +777,142 @@ def add_extra_tokens_for_images_gemma3(
   )
 
 
+def add_extra_tokens_for_qwen3_omni(
```
Comment (hengtaoguo) [inline src/MaxText/multimodal_utils.py]: Looks like this function has not yet been added to the above `prepare_text_for_image_fusion` (which is OK for this PR). Can you move this `add_extra_tokens_for_qwen3_omni` to [multimodal/qwen3_omni_processor.py](https://github.com/AI-Hypercomputer/maxtext/blob/main/src/MaxText/multimodal/qwen3_omni_processor.py)?

File: src/maxtext/decode.py [hunk 2]
```diff
@@ -120,12 +120,29 @@ def main(argv: Sequence[str]) -> None:
...
+  position_ids = None
+  mrope_position_deltas = None
+
   if config.use_multimodal:
     tokens = multimodal_utils.prepare_text_for_image_fusion(
         tokens, model_name=config.model_name, processor_output=processor_outputs
     )
     true_length += image_offsets
 
+    if config.use_mrope:
```
Comment (aireenmei) [inline src/maxtext/decode.py:143]: Should we infer the requirement of mrope from the modalities in processor_outputs instead of using a flag use_mrope? To avoid breakage if the user input multimodal data but set use_mrope=false.

File: src/MaxText/input_pipeline/_input_pipeline_utils.py [hunk 3]
```diff
@@ -712,3 +713,105 @@ def __init__(self, ignored_ids, axis=1):
...
+    Returns:
+      element with {data_column}_position updated to shape (3, batch, seq_len)
+      for 3D positions (always 3D, even for text-only sequences).
+    """
+
+    # Extract inputs and metadata
+    input_ids = element[self.data_column]
+    attention_mask = element.get(f"{self.data_column}_segmentation")
+
+    # Convert to JAX arrays if needed
```
Comment (aireenmei) [inline src/MaxText/input_pipeline/_input_pipeline_utils.py]: Since the data input pipeline is running on CPU, we want to use numpy instead of JAX to avoid any conflicts. I see we do these conversions because multimodal_utils.get_rope_index expects JAX Array and use jnp. Could you rewrite multimodal_utils.get_rope_index to use numpy and remove the JAX Array conversions?

File: tests/unit/multimodal_rope_check.py [hunk 4]
```diff
@@ -0,0 +1,675 @@
...
+      use_audio_in_video: Whether to interleave audio with video
+      audio_lengths: Optional (num_audios,)
+      second_per_grids: Optional (num_videos,)
+      spatial_merge_size: Spatial merge size
+      position_id_per_seconds: Temporal granularity
+
+    Returns:
+      Tuple of (jax_position_ids, pytorch_position_ids, match_status)
+    """
+    # JAX version
```
Comment (aireenmei) [inline tests/unit/multimodal_rope_check.py]: Use numpy instead of jnp.

File: src/MaxText/layers/embeddings.py [hunk 5]
```diff
@@ -1539,3 +1539,194 @@ def __call__(self, num_frames: int, height: int, width: int) -> Array:
       interpolated = interpolated.astype(self.fprop_dtype)
 
     return interpolated
+
+
+class Qwen3OmniMoeThinkerTextRotaryEmbedding(nnx.Module):
```
Comment (aireenmei) [inline src/MaxText/layers/embeddings.py]: How about inheriting from RotaryEmbedding to reuse code for calculating frequencies. Gemini suggests this way: https://paste.googleplex.com/5160930042314752

**Comment by eitanporat** [comment]
If you need help hit me up 😀

File: src/maxtext/decode.py (hunk 2 above)
Comment (entrpn) [reply src/maxtext/decode.py:143]: No because not all multimodal models use mrope.

File: src/MaxText/input_pipeline/_input_pipeline_utils.py (hunk 3 above)
Comment (entrpn) [reply src/MaxText/input_pipeline/_input_pipeline_utils.py]: get_rope_index is called in decode.py. What about using `with jax.default_device(cpu_device):` in this pipeline only and leaving the implementation in JAX?

File: src/MaxText/multimodal_utils.py (hunk 1 above)
Comment (entrpn) [reply src/MaxText/multimodal_utils.py]: done

File: src/MaxText/layers/embeddings.py (hunk 5 above)
Comment (entrpn) [reply src/MaxText/layers/embeddings.py]: done.

**Comment by entrpn** [comment]
> If you need help hit me up 😀

Thanks Eitan, will reach out if I have any questions.

File: src/MaxText/input_pipeline/_input_pipeline_utils.py (hunk 3 above)
Comment (hengtaoguo) [reply src/MaxText/input_pipeline/_input_pipeline_utils.py]: I remember we did the same `with jax.default_device(cpu_device)` for image-batch-resize in preprocessing utils, but it will still occupy the TPU during SFT? Then I rewrote that using PIL libs. @aireenmei

Comment (entrpn) [reply src/MaxText/input_pipeline/_input_pipeline_utils.py]: I will write with numpy.

File: src/MaxText/layers/embeddings.py [hunk 6]
```diff
@@ -1539,3 +1546,180 @@ def __call__(self, num_frames: int, height: int, width: int) -> Array:
       interpolated = interpolated.astype(self.fprop_dtype)
 
     return interpolated
+
+
+class Qwen3OmniMoeThinkerTextRotaryEmbedding(RotaryEmbedding):
```
Comment (gagika) [inline src/MaxText/layers/embeddings.py:1554]: seams this file is getting too long, could you consider splitting it?

Comment (entrpn) [reply src/MaxText/layers/embeddings.py:1554]: @hengtaoguo is restructuring some of these classes in another PR, I will work with him on this in that PR.

**Review by gagika — APPROVED** [review]

**Review by aireenmei — APPROVED** [review]
